security/pkg/stsservice/tokenmanagers/google/tokenmanager.go

```diff
@@ -0,0 +1,303 @@
+// Example of a federated token request:
+// POST https://securetoken.googleapis.com/v1/identitybindingtoken
+// Content-Type: application/json
+// {
+//    audience: <trust domain>
+//    grantType: urn:ietf:params:oauth:grant-type:token-exchange
+//    requestedTokenType: urn:ietf:params:oauth:token-type:access_token
+//    subjectTokenType: urn:ietf:params:oauth:token-type:jwt
+//    subjectToken: <jwt token>
+//    Scope: https://www.googleapis.com/auth/cloud-platform
+// }
+func (tm *TokenManager) constructFederatedTokenRequest(parameters stsservice.StsRequestParameters) *http.Request {
+	reqScope := scope
+	if len(parameters.Scope) != 0 {
+		reqScope = parameters.Scope
+	}
+	query := map[string]string{
+		"audience":           tm.trustDomain,
+		"grantType":          parameters.GrantType,
+		"requestedTokenType": tokenType,
+		"subjectTokenType":   parameters.SubjectTokenType,
+		"subjectToken":       parameters.SubjectToken,
+		"Scope":              reqScope,
+	}
+	jsonQuery, _ := json.Marshal(query)
+	req, _ := http.NewRequest("POST", federatedTokenEndpoint, bytes.NewBuffer(jsonQuery))
+	req.Header.Set("Content-Type", contentType)
+	reqDump, _ := httputil.DumpRequest(req, true)
+	tokenManagerLog.Debugf("Generated federated token request: %s", string(reqDump))
+	return req
+}
+
+// fetchFederatedToken exchanges a third-party issued Json Web Token for an OAuth2.0 access token
+// which asserts a third-party identity within an identity namespace.
+func (tm *TokenManager) fetchFederatedToken(parameters stsservice.StsRequestParameters) (*federatedTokenResponse, error) {
+	respData := &federatedTokenResponse{}
+
+	req := tm.constructFederatedTokenRequest(parameters)
+	resp, err := tm.sendRequestWithRetry(req)
+	if err != nil {
+		respCode := 0
+		if resp != nil {
+			respCode = resp.StatusCode
+		}
+		tokenManagerLog.Errorf("Failed to exchange federated token (HTTP status %d): %v", respCode,
+			err)
+		return nil, fmt.Errorf("failed to exchange federated token (HTTP status %d): %v", respCode,
+			err)
+	}
+	// resp should not be nil.
+	defer resp.Body.Close()
+
+	respDump, _ := httputil.DumpResponse(resp, true)
+	tokenManagerLog.Debugf("Received federated token response: \n%s", string(respDump))
+
+	body, _ := ioutil.ReadAll(resp.Body)
+	if err := json.Unmarshal(body, respData); err != nil {
+		tokenManagerLog.Errorf("Failed to unmarshal federated token response data: %v", err)
+		return respData, fmt.Errorf("failed to unmarshal federated token response data: %v", err)
+	}
+	if respData.AccessToken == "" {
+		tokenManagerLog.Errora("federated token response does not have access token", string(body))
+		return respData, errors.New("federated token response does not have access token. " + string(body))
+	}
+	tokenManagerLog.Debug("successfully exchanged a federated token")
+	tokenReceivedTime := time.Now()
+	tm.tokens.Store(tokenReceivedTime.String(), stsservice.TokenInfo{
```

Comment:
Where is "tokens" used? Is it for debugging purpose only?

Reply:
Yes

Reply:
Yes